2025-03-03 ZeroTier - health checking - alternative proposal

Paraphraser opened the pull request:
This PR follows on from the extensive discussion associated with #37.

Never before have I even *contemplated* submitting a PR covering the same ground as an existing open PR. However, on this occasion I thought it might be useful to have a concrete proposal to compare and contrast with #37.

I sincerely hope that laying this on the (virtual) table and then minimising further interaction *might* help us converge on a solution.

<hr>

* `docker-compose.yml` and `docker-compose-router.yml`:

	- replaces deprecated `version` statement with `---`.

	- adds example environment variables.

* `Dockerfile`

	- corrects case of "as" to "AS" (silences build warning).

	- adds and configures `healthcheck.sh` (as per #37).

	- includes `tzdata` package (moved from `Dockerfile.router`) so messages have local timestamps.

* `Dockerfile.router`

	- removes `tzdata` (moved to `Dockerfile`).

* `entrypoint-router.sh`:

	- code for first launch auto join of listed networks expanded to include additional help material.

* `entrypoint.sh`:

	- "first launch" auto join of listed networks (code copied from `entrypoint-router.sh`, as modified per above).

	- "self repair" of permissions in persistent store (code copied from `entrypoint-router.sh`).

	- adds launch-time message to make it clear that the client is launching (complements messages in `entrypoint-router.sh`).

	- abstracts some common strings to environment variables (opportunistic change).

* `README.md`:

	- updates examples.

	- describes new environment variables (including move of `ZEROTIER_ONE_NETWORK_IDS` from `README-router.md`.

	- documents health-checking.

* `README-router.md`

	- updates examples.
	- explains relationship of router and client.

Added:

* `healthcheck.sh`, based on original proposal in #37 and subsequent suggestions for modification by me.

I gave serious consideration to the code for synchronising networks in the entry point scripts. The idea is quite attractive. It is safe to automate joins in a "clean slate" situation. However, a *leave* followed by a *join* is not guaranteed to be idempotent. That's because the *leave* destroys the network-specific configuration options (`allowManaged`, `allowGlobal`, `allowDefault`, `allowDNS`).

On balance I think it's better left to users to send explicit *leave* commands via the CLI and take responsibility for restoring lost configuration options on any subsequent *join*.

I will post the results of testing this PR separately.

<hr>

Additional changes as at 2025-04-06

`healthcheck.sh`:

1. Full rewrite, including copious comments explaining theory of
   operation. In essence, if a «networkID» is mentioned in
   (internal path):

	```
	/var/lib/zerotier-one/networks.d
	```

	then it should be matched by a route in the host's routing table:
	
	* Zero «networkID» = zero routes
	* One «networkID» = one route
	* Two «networkID» = two routes
	* ...

	Any mismatch causes the container to go unhealthy. From the
	perspective of the health-checking script, the question of *which*
	network is immaterial so there is no need to employ techniques such
	as iterating to discover *which* network is causing a problem. This
	is because the script has no way of communicating anything other
	than an exit status. Any `echo` statements it issues will not make
	it into the container's log.
	
	The simple presence/absence of a «networkID» in `networks.d` is
	taken to indicate which networks the user *intends* the container
	to join.
	
	There is no reliance on environment variables to propagate any
	health-checking information into the container. No other variables
	are introduced so the argument about naming conventions goes away.

`Dockerfile`:

1. Removes `--start-interval` flag from `HEALTHCHECK` command. The flag
   was preventing `buildah` builds from succeeding.

`README.md`:

1. Removes references to the following environment variables:

	* `ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`
	* `ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH`

2. Rewrites explanation of health-checking.

`README-router.md`:

1. Removes references to the following environment variables:

	* `ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`
	* `ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH`

Paraphraser committed:
2025-03-03 ZeroTier - health checking - alternative proposal

This PR follows on from the extensive discussion associated with #37.

Never before have I even *contemplated* submitting a PR covering the
same ground as an existing open PR. However, on this occasion I thought
it might be useful to have a concrete proposal to compare and contrast
with #37.

I sincerely hope that laying this on the (virtual) table and then
minimising further interaction *might* help us converge on a solution.

<hr>

Changes:

* `docker-compose.yml` and `docker-compose-router.yml`:

	- replaces deprecated `version` statement with `---`.

	- adds example environment variables.

* `Dockerfile`

	- corrects case of "as" to "AS" (silences build warning).

	- adds and configures `healthcheck.sh` (as per #37).

	- includes `tzdata` package (moved from `Dockerfile.router`) so
	  messages have local timestamps.

* `Dockerfile.router`

	- removes `tzdata` (moved to `Dockerfile`).

* `entrypoint-router.sh`:

	- code for first launch auto join of listed networks expanded to
	  include additional help material.

* `entrypoint.sh`:

	- "first launch" auto join of listed networks (code copied from
	  `entrypoint-router.sh`, as modified per above).

	- "self repair" of permissions in persistent store (code copied from
	  `entrypoint-router.sh`).

	- adds launch-time message to make it clear that the client is
	  launching (complements messages in `entrypoint-router.sh`).

	- abstracts some common strings to environment variables
	  (opportunistic change).

* `README.md`:

	- updates examples.

	- describes new environment variables (including move of
	  `ZEROTIER_ONE_NETWORK_IDS` from `README-router.md`.

	- documents health-checking.

* `README-router.md`

	- updates examples.
	- explains relationship of router and client.

Added:

* `healthcheck.sh`, based on original proposal in #37 and subsequent
  suggestions for modification by me.

I gave serious consideration to the code for synchronising networks in
the entry point scripts. The idea is quite attractive. It is safe to
automate joins in a "clean slate" situation. However, a *leave* followed
by a *join* is not guaranteed to be idempotent. That's because the
*leave* destroys the network-specific configuration options
(`allowManaged`, `allowGlobal`, `allowDefault`, `allowDNS`).

On balance I think it's better left to users to send explicit *leave*
commands via the CLI and take responsibility for restoring lost
configuration options on any subsequent *join*.

I will post the results of testing this PR separately.

Signed-off-by: Phill Kelley <34226495+Paraphraser@users.noreply.github.com>

Paraphraser commented:
# Testing

This post includes a series of tests I have run against containers built using the changes proposed in this PR.

## reference service definitions

### client

``` yaml
  zerotier:
    container_name: zerotier
    image: "zyclonite/zerotier:local"
    restart: unless-stopped
    network_mode: host
    environment:
      - TZ=${TZ:-Etc/UTC}
      # - ZEROTIER_ONE_NETWORK_IDS=9999888877776666
      # - ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS=9999888877776666
      # - ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH=1
    volumes:
      - ./volumes/zerotier-one:/var/lib/zerotier-one
    devices:
      - "/dev/net/tun:/dev/net/tun"
    cap_add:
      - NET_ADMIN
      - SYS_ADMIN
```

### router

``` yaml
  zerotier-router:
    container_name: zerotier
    image: "zyclonite/zerotier-router:local"
    restart: unless-stopped
    network_mode: host
    environment:
      - TZ=${TZ:-Etc/UTC}
      - ZEROTIER_ONE_LOCAL_PHYS=ens18
      - ZEROTIER_ONE_USE_IPTABLES_NFT=true
      - ZEROTIER_ONE_GATEWAY_MODE=both
      # - ZEROTIER_ONE_NETWORK_IDS=9999888877776666
      # - ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS=9999888877776666
      # - ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH=1
    volumes:
      - ./volumes/zerotier-one:/var/lib/zerotier-one
    devices:
      - "/dev/net/tun:/dev/net/tun"
    cap_add:
      - NET_ADMIN
      - SYS_ADMIN
      - NET_RAW
```

## client testing

### clean slate - null configuration

#### configuration

environment variable                     | defined
-----------------------------------------|----------
`ZEROTIER_ONE_NETWORK_IDS`               | no
`ZEROTIER_ONE_LOCAL_PHYS`                | no
`ZEROTIER_ONE_USE_IPTABLES_NFT`          | no
`ZEROTIER_ONE_GATEWAY_MODE`              | no
`ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH` | no<sup>1</sup>
`ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`     | no

1.	Defaults to "1". Takes effect because specific network check is undefined.

#### setup

```
$ ls ./volumes/zerotier-one
ls: cannot access './volumes/zerotier-one': No such file or directory

$ docker-compose up -d zerotier
[+] Running 1/1
 ✔ Container zerotier  Started                                                                                         0.2s 

$ docker logs zerotier
Sun Mar  2 23:34:37 AEDT 2025 - assuming container first run.
 ZEROTIER_ONE_NETWORK_IDS not set. You will need to join
 networks using zerotier-cli, and then approve this
 host in ZeroTier Central.
Sun Mar  2 23:34:37 AEDT 2025 - launching ZeroTier-One in client mode
Starting Control Plane...
Starting V6 Control Plane...

$ docker exec zerotier env | grep '^ZEROTIER_ONE_'
```

Interpretation:

1. Show persistent store does not exist.
2. Container launched.
3. Log confirms client and reminds user to join a network and approve this host.
4. No variables defined.

#### test

```
$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                     PORTS     NAMES
6f51e4e8a18c   zyclonite/zerotier:local   "entrypoint.sh -U"   2 minutes ago   Up 2 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>

$ docker exec zerotier zerotier-cli join 9999888877776666
200 join OK

$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 9999888877776666  22:ea:af:dc:d7:94 REQUESTING_CONFIGURATION PRIVATE ztr2qsmswx -

$ docker exec zerotier zerotier-cli get 9999888877776666 status
REQUESTING_CONFIGURATION

$ ip r | grep 'dev zt' | grep -cv via
0

$ echo 'Go to ZeroTier Central and approve this client'
Go to ZeroTier Central and approve this client

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                   PORTS     NAMES
6f51e4e8a18c   zyclonite/zerotier:local   "entrypoint.sh -U"   3 minutes ago   Up 3 minutes (healthy)             zerotier

$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 9999888877776666 My_ZeroTier 22:ea:af:dc:d7:94 OK PRIVATE ztr2qsmswx 10.244.63.140/16

$ docker exec zerotier zerotier-cli get 9999888877776666 status
OK

$ ip r | grep 'dev zt' | grep -cv via
1

$ docker-compose down zerotier
[+] Running 1/1
 ✔ Container zerotier  Removed                                                                                         2.6s 
```

Interpretation:

1. Show container unhealthy (waiting for a join). Expected result.
2. Show container has not joined any networks.
3. Join a network.
4. Show `listnetworks` status is `REQUESTING_CONFIGURATION`.
5. Show status of joined network is `REQUESTING_CONFIGURATION`.
6. Show no routes associated with ZeroTier interface (because interface is not constructed until client is approved).
7. Approve client (via browser).
8. Show container goes healthy. Expected result.
9. Show `listnetworks` status is `OK`.
10. Show status of joined network is `OK`.
11. Show direct route to ZeroTier interface has been created.
12. End of test. 

### minimum route checking

#### configuration

environment variable                     | defined
-----------------------------------------|----------
`ZEROTIER_ONE_NETWORK_IDS`               | yes
`ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH` | no<sup>1</sup>
`ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`     | no

1.	Defaults to "1". Takes effect because specific network check is undefined.

#### setup

```
$ docker-compose up -d zerotier
[+] Running 1/1
 ✔ Container zerotier  Started                                                                                                                             0.2s 

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                   PORTS     NAMES
656c6c75c8f3   zyclonite/zerotier:local   "entrypoint.sh -U"   5 seconds ago   Up 5 seconds (healthy)             zerotier

$ docker logs zerotier
Wed Feb 26 14:20:51 AEDT 2025 - launching ZeroTier-One in client mode
Starting Control Plane...
Starting V6 Control Plane...

$ docker exec zerotier env | grep '^ZEROTIER_ONE_'
ZEROTIER_ONE_NETWORK_IDS=9999888877776666
```

Interpretation:

1. Container launched.
2. Container healthy. Expected result.
3. Log confirms client.
4. Only one variable defined.

#### test

```
$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 9999888877776666 My_ZeroTier 22:43:c0:8e:8d:1c OK PRIVATE ztr2qsmswx 10.244.211.241/16

$ docker exec zerotier zerotier-cli leave 9999888877776666
200 leave OK

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                     PORTS     NAMES
656c6c75c8f3   zyclonite/zerotier:local   "entrypoint.sh -U"   2 minutes ago   Up 2 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli join 9999888877776666
200 join OK

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                   PORTS     NAMES
656c6c75c8f3   zyclonite/zerotier:local   "entrypoint.sh -U"   3 minutes ago   Up 3 minutes (healthy)             zerotier

$ sudo nmcli conn down ztr2qsmswx
Connection 'ztr2qsmswx' successfully deactivated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/86)

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                     PORTS     NAMES
656c6c75c8f3   zyclonite/zerotier:local   "entrypoint.sh -U"   5 minutes ago   Up 5 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli get 9999888877776666 status
OK

$ ip r | grep 'dev zt' | grep -cv via
0

$ docker-compose down zerotier
[+] Running 1/1
 ✔ Container zerotier  Removed                                                                                                                             1.4s 
```

Interpretation:

1. Show container has joined test network.
2. Leave that network.
3. Show container goes unhealthy. Expected result.
4. Rejoin the network.
5. Show container goes healthy. Expected result.
6. Destroy the network interface.
7. Show container goes unhealthy. Expected result.
8. Show reason is **not** because the ZeroTier network has gone away. Expected result.
9. Show reason is **because** the direct route has gone away (count is zero).  Expected result.
10. End of test. 

### specific network checking

#### configuration

environment variable                     | defined
-----------------------------------------|----------
`ZEROTIER_ONE_NETWORK_IDS`               | yes
`ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH` | no
`ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`     | yes<sup>1</sup>

1. Takes precedence over default of 1 for minimum route checking.

#### setup

```
$ docker-compose up -d zerotier
[+] Running 1/1
 ✔ Container zerotier  Started                                                                                                                                     0.2s 

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                   PORTS     NAMES
6d1e7c9d155a   zyclonite/zerotier:local   "entrypoint.sh -U"   5 seconds ago   Up 5 seconds (healthy)             zerotier

$ docker logs zerotier
Wed Feb 26 14:44:12 AEDT 2025 - launching ZeroTier-One in client mode
Starting Control Plane...
Starting V6 Control Plane...

$ docker exec zerotier env | grep '^ZEROTIER_ONE_'
ZEROTIER_ONE_NETWORK_IDS=9999888877776666
ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS=9999888877776666
```

Interpretation:

1. Container launched.
2. Container healthy. Expected result.
3. Log confirms client.
4. Two variables defined.

#### test

```
$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 9999888877776666 My_ZeroTier 22:43:c0:8e:8d:1c OK PRIVATE ztr2qsmswx 10.244.211.241/16

$ docker exec zerotier zerotier-cli leave 9999888877776666
200 leave OK

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                     PORTS     NAMES
6d1e7c9d155a   zyclonite/zerotier:local   "entrypoint.sh -U"   2 minutes ago   Up 2 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli join 9999888877776666
200 join OK

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                   PORTS     NAMES
6d1e7c9d155a   zyclonite/zerotier:local   "entrypoint.sh -U"   3 minutes ago   Up 3 minutes (healthy)             zerotier

$ sudo nmcli conn down ztr2qsmswx
Connection 'ztr2qsmswx' successfully deactivated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/88)

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                     PORTS     NAMES
6d1e7c9d155a   zyclonite/zerotier:local   "entrypoint.sh -U"   5 minutes ago   Up 5 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli get 9999888877776666 status
OK

$ ip r | grep 'dev zt' | grep -cv via
0

$ docker-compose down zerotier
[+] Running 1/1
 ✔ Container zerotier  Removed                                                                                                                                     1.3s 
```

Interpretation:

1. Show container has joined test network.
2. Leave that network.
3. Show container goes unhealthy. Expected result.
4. Rejoin the network.
5. Show container goes healthy. Expected result.
6. Destroy the network interface.
7. Show container goes unhealthy. Expected result.
8. Show reason is **not** because the ZeroTier network has gone away. Expected result.
9. Show reason is **because** the direct route has gone away (count is zero). Expected result.
10. End of test. 

<hr>

## router testing

### minimum route checking

#### configuration

environment variable                     | defined
-----------------------------------------|----------
`ZEROTIER_ONE_NETWORK_IDS`               | yes
`ZEROTIER_ONE_LOCAL_PHYS`                | yes
`ZEROTIER_ONE_USE_IPTABLES_NFT`          | yes
`ZEROTIER_ONE_GATEWAY_MODE`              | yes
`ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH` | no<sup>1</sup>
`ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`     | no

1.	Defaults to "1". Takes effect because specific network check is undefined.

#### setup

```
$ docker-compose up -d zerotier-router
[+] Running 1/1
 ✔ Container zerotier  Started                                                                                                                                     0.2s 

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                   PORTS     NAMES
f23bcc78f936   zyclonite/zerotier-router:local   "entrypoint-router.s…"   6 seconds ago   Up 5 seconds (healthy)             zerotier

$ docker logs zerotier
Wed Feb 26 15:02:36 AEDT 2025 - launching ZeroTier-One in routing mode
adding iptables-nft rules for bi-directional traffic (local interfaces ens18 to/from ZeroTier)
Wed Feb 26 15:02:36 AEDT 2025 - ZeroTier daemon is running as process 17
Starting Control Plane...
Starting V6 Control Plane...

$ docker exec zerotier env | grep '^ZEROTIER_ONE_'
ZEROTIER_ONE_NETWORK_IDS=9999888877776666
ZEROTIER_ONE_LOCAL_PHYS=ens18
ZEROTIER_ONE_USE_IPTABLES_NFT=true
ZEROTIER_ONE_GATEWAY_MODE=both
```

Interpretation:

1. Container launched.
2. Container healthy. Expected result.
3. Log confirms router.
4. Four variables defined.

#### test

```
$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 9999888877776666 My_ZeroTier 22:43:c0:8e:8d:1c OK PRIVATE ztr2qsmswx 10.244.211.241/16

$ docker exec zerotier zerotier-cli leave 9999888877776666
200 leave OK

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                     PORTS     NAMES
f23bcc78f936   zyclonite/zerotier-router:local   "entrypoint-router.s…"   2 minutes ago   Up 2 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli join 9999888877776666
200 join OK

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                   PORTS     NAMES
f23bcc78f936   zyclonite/zerotier-router:local   "entrypoint-router.s…"   3 minutes ago   Up 3 minutes (healthy)             zerotier

$ sudo nmcli conn down ztr2qsmswx
Connection 'ztr2qsmswx' successfully deactivated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/91)

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                     PORTS     NAMES
f23bcc78f936   zyclonite/zerotier-router:local   "entrypoint-router.s…"   5 minutes ago   Up 5 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli get 9999888877776666 status
OK

$ ip r | grep 'dev zt' | grep -cv via
0

$ docker-compose down zerotier-router
[+] Running 1/1
 ✔ Container zerotier  Removed                                                                                                                                     1.4s 
```

Interpretation:

1. Show container has joined test network.
2. Leave that network.
3. Show container goes unhealthy. Expected result.
4. Rejoin the network.
5. Show container goes healthy. Expected result.
6. Destroy the network interface.
7. Show container goes unhealthy. Expected result.
8. Show reason is **not** because the ZeroTier network has gone away. Expected result.
9. Show reason is **because** the direct route has gone away (count is zero). Expected result.
10. End of test. 

### specific network checking

#### configuration

environment variable                     | defined
-----------------------------------------|----------
`ZEROTIER_ONE_NETWORK_IDS`               | yes
`ZEROTIER_ONE_LOCAL_PHYS`                | yes
`ZEROTIER_ONE_USE_IPTABLES_NFT`          | yes
`ZEROTIER_ONE_GATEWAY_MODE`              | yes
`ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH` | no
`ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`     | yes<sup>1</sup>

1. Takes precedence over default of 1 for minimum route checking.

#### setup

```
$ docker-compose up -d zerotier-router
[+] Running 1/1
 ✔ Container zerotier  Started                                                                                                                                     0.2s 

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                   PORTS     NAMES
ea0f81101029   zyclonite/zerotier-router:local   "entrypoint-router.s…"   5 seconds ago   Up 5 seconds (healthy)             zerotier

$ docker logs zerotier
Wed Feb 26 15:26:08 AEDT 2025 - launching ZeroTier-One in routing mode
adding iptables-nft rules for bi-directional traffic (local interfaces ens18 to/from ZeroTier)
Wed Feb 26 15:26:08 AEDT 2025 - ZeroTier daemon is running as process 17
Starting Control Plane...
Starting V6 Control Plane...

$ docker exec zerotier env | grep '^ZEROTIER_ONE_'
ZEROTIER_ONE_LOCAL_PHYS=ens18
ZEROTIER_ONE_USE_IPTABLES_NFT=true
ZEROTIER_ONE_GATEWAY_MODE=both
ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS=9999888877776666
ZEROTIER_ONE_NETWORK_IDS=9999888877776666
```

Interpretation:

1. Container launched.
2. Container healthy. Expected result.
3. Log confirms router.
4. Five variables defined.

#### test

```
$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 9999888877776666 My_ZeroTier 22:43:c0:8e:8d:1c OK PRIVATE ztr2qsmswx 10.244.211.241/16

$ docker exec zerotier zerotier-cli leave 9999888877776666
200 leave OK

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                     PORTS     NAMES
ea0f81101029   zyclonite/zerotier-router:local   "entrypoint-router.s…"   2 minutes ago   Up 2 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli join 9999888877776666
200 join OK

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                   PORTS     NAMES
ea0f81101029   zyclonite/zerotier-router:local   "entrypoint-router.s…"   3 minutes ago   Up 3 minutes (healthy)             zerotier

$ sudo nmcli conn down ztr2qsmswx
Connection 'ztr2qsmswx' successfully deactivated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/95)

$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED         STATUS                     PORTS     NAMES
ea0f81101029   zyclonite/zerotier-router:local   "entrypoint-router.s…"   5 minutes ago   Up 5 minutes (unhealthy)             zerotier

$ docker exec zerotier zerotier-cli get 9999888877776666 status
OK

$ ip r | grep 'dev zt' |grep -cv via
0

+ docker-compose down zerotier-router
[+] Running 1/1
 ✔ Container zerotier  Removed                                                                                                                                     1.5s 
```

Interpretation:

1. Show container has joined test network.
2. Leave that network.
3. Show container goes unhealthy. Expected result.
4. Rejoin the network.
5. Show container goes healthy. Expected result.
6. Destroy the network interface.
7. Show container goes unhealthy. Expected result.
8. Show reason is **not** because the ZeroTier network has gone away. Expected result.
9. Show reason is **because** the direct route has gone away (count is zero). Expected result.
10. End of test.

Paraphraser commented:
Hmmm. That's a bit odd:

![Screenshot 2025-03-03 at 14 05 47](https://github.com/user-attachments/assets/1aea42f2-a6fe-4c31-a0a9-38979b31749e)

Seems to be moaning about the --start-interval flag but that's definitely defined in the [Dockerfile documentation](https://docs.docker.com/reference/dockerfile/#healthcheck).

I don't get this error when I do a local build. Any ideas?

gb-123-git commented:
@Paraphraser @zyclonite 

My humble question for a small use case scenario :
What if the user wants to check ALL networks he has joined (which can change dynamically) ?
How do we check that in this proposal ?

Paraphraser commented:
@zyclonite - Lukas does [this comment](https://github.com/containers/buildah/issues/6019#issuecomment-2695765812) make any sense to you in the ZeroTier build context?

I have no experience with GitHub Actions so I don't understand anything about how these things run. Do **you** get to specify the version of buildah, or is this just a service you're using that's maintained by someone else (eg GitHub provides this as a standard service)?

According to [buildah install.md](https://github.com/containers/buildah/blob/main/install.md), for Debian I need Bookworm plus a standard `apt install buildah`. On Bookworm 12.9 that gets me `buildah version 1.28.2`.

For Ubuntu the prerequisite is 20.10 and newer, also with a standard `apt install`. On Ubuntu 24.04.2 that gets me `buildah version 1.33.7` which produces the same error about `--start-interval`.

Both of these versions are less than the [v1.37.0](https://github.com/containers/buildah/releases/tag/v1.37.0) cited as where the fix went into production.

The latest and greatest appears to be 1.39.1.

I don't get why the Bookworm and Noble Numbat `apt` installs are so out-of-date.

I started to read [Building from scratch](https://github.com/containers/buildah/blob/main/install.md#building-from-scratch) but my eyes quickly glazed over.

zyclonite commented:
first i was on vacation and last week a bit downing in work - sorry for the delay

regarding buildah version - i fear we have to live with the one provided with the ubuntu 24 gh actions runner (they might upgrade it from time to time)
i did attempt to upgrade it individually in the past but they have quite some security hurdles around that, so it's not easy to achieve (this was my old project but it does not work anymore with the latest runners https://github.com/zyclonite/setup-podman)

about the sponsored by comment - i fully support individual contributions, so you are free to add your real name but i would like to not go down the route with having companies sponsoring code as this might be tricky from a licensing perspective if not in sync with the individual contributor and so on...

gb-123-git commented:
@zyclonite 
After due discussions with PMGA Tech LLP, I have changed the License text to match most major MIT license texts:
E.g.
**TailWindCSS**
https://github.com/tailwindlabs/tailwindcss/blob/main/LICENSE
**VS-Code** (By Microsoft)
https://github.com/microsoft/vscode/blob/main/LICENSE.txt
This should alleviate your fears as the text is matching *most* major MIT projects.

However, as per my agreement with PMGA Tech LLP, the code **_cannot_** be used without attaching the proper licensing text. It is now not possible for me to go back on it.

Further, I still do not understand the fuss behind all this as many major MIT projects owned by corporates also include license files (Examples already given above); and also since ZeroTier itself is **_not_** MIT but BSL, you can check here: https://github.com/zerotier/ZeroTierOne?tab=License-1-ov-file 

As I do not want to take this argument further, so its your call on whether to merge the code or cancel it. 

Please note one more thing, in case you intend to merge this proposal, the proper licensing text needs to be copied from PR #37 .
I'm afraid if the text is not attached 'as-is', it would be a deliberate copywrite infringement since knowingly the same has been removed and/or not attached.
  
I would suggest you merge PR #37 since this is just a copy of that MR with minor changes as already admitted by the OP in the first post. (attaching screenshot for future reference in case required)

![Derivative Proof](https://github.com/user-attachments/assets/fc500e4e-f328-4f9a-bc9e-703fa2ed6d70)

Paraphraser committed:
`healthcheck.sh`:

1. Full rewrite, including copious comments explaining theory of
   operation. In essence, if a «networkID» is mentioned in
   (internal path):

	```
	/var/lib/zerotier-one/networks.d
	```

	then it should be matched by a route in the host's routing table:

	* Zero «networkID» = zero routes
	* One «networkID» = one route
	* Two «networkID» = two routes
	* ...

	Any mismatch causes the container to go unhealthy. From the
	perspective of the health-checking script, the question of *which*
	network is immaterial so there is no need to employ techniques such
	as iterating to discover *which* network is causing a problem. This
	is because the script has no way of communicating anything other
	than an exit status. Any `echo` statements it issues will not make
	it into the container's log.

	The simple presence/absence of a «networkID» in `networks.d` is
	taken to indicate which networks the user *intends* the container
	to join.

	There is no reliance on environment variables to propagate any
	health-checking information into the container. No other variables
	are introduced so the argument about naming conventions goes away.

`Dockerfile`:

1. Removes `--start-interval` flag from `HEALTHCHECK` command. The flag
   was preventing `buildah` builds from succeeding.

`README.md`:

1. Removes references to the following environment variables:

	* `ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`
	* `ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH`

2. Rewrites explanation of health-checking.

`README-router.md`:

1. Removes references to the following environment variables:

	* `ZEROTIER_ONE_CHK_SPECIFIC_NETWORKS`
	* `ZEROTIER_ONE_CHK_MIN_ROUTES_FOR_HEALTH`

Signed-off-by: Phill Kelley <34226495+Paraphraser@users.noreply.github.com>

Paraphraser commented:
# Testing (with 2025-04-06 changes)

## Reference service definition

``` yaml
  zerotier:
    container_name: zerotier
    image: "zyclonite/zerotier:local"
    restart: unless-stopped
    network_mode: host
    environment:
      - TZ=${TZ:-Etc/UTC}
    # - ZEROTIER_ONE_NETWORK_IDS=${ZEROTIER_ONE_NETWORK_IDS}
    volumes:
      - ./volumes/zerotier-one:/var/lib/zerotier-one
    devices:
      - "/dev/net/tun:/dev/net/tun"
    cap_add:
      - NET_ADMIN
      - SYS_ADMIN
```

Note:

* `ZEROTIER_ONE_NETWORK_IDS` commented-out to disable auto-join in clean-slate situation.

## Test 1 - clean slate

1. Show container not running and no persistent store:

	``` console
	$ docker ps
	CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
	
	$ ls -ld ~/IOTstack/volumes/zerotier-one
	ls: cannot access '/home/moi/IOTstack/volumes/zerotier-one': No such file or directory
	```

2. Start container

	``` console
	$ docker compose up -d zerotier
	```
	
	Show persistent store created:
	
	``` console
	$ ls -ld ~/IOTstack/volumes/zerotier-one
	drwxr-xr-x 4 999 994 4096 Apr  6 09:46 /home/moi/IOTstack/volumes/zerotier-one
	```
	
	Show networks directory does not exist:
	
	``` console
	$ ls -ld ~/IOTstack/volumes/zerotier-one/networks.d
	ls: cannot access '/home/moi/IOTstack/volumes/zerotier-one/networks.d': No such file or directory
	```
	
	Show container healthy:
	
	``` container
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                   PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   3 minutes ago   Up 3 minutes (healthy)             zerotier
	```
	
## Join first network

1. Join network:

	``` console
	$ docker exec zerotier zerotier-cli join 9999888877776666
	200 join OK
	```

	Show container goes unhealthy:
	
	```
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                     PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   6 minutes ago   Up 6 minutes (unhealthy)             zerotier
	```
	
	Explore reason:
	
	```
	$ docker exec zerotier zerotier-cli listnetworks
	200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
	200 listnetworks 9999888877776666  22:ef:5f:10:91:a9 ACCESS_DENIED PRIVATE ztr2qsmswx -
	
	$ docker exec zerotier zerotier-cli get 9999888877776666 status
	ACCESS_DENIED
	```

	Authorise client in ZeroTier Central. Then show container goes healthy.
	
	``` console
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED          STATUS                    PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   10 minutes ago   Up 10 minutes (healthy)             zerotier
	```

## Test 3 - interrupt first network
	
1. List networks:

	``` console
	$ ip r | grep "dev zt.* scope link"
	10.244.0.0/16 dev ztr2qsmswx proto kernel scope link src 10.244.235.233 
	```
	
2. Destroy network:

	``` console
	$ sudo nmcli conn down ztr2qsmswx
	Connection 'ztr2qsmswx' successfully deactivated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/144)
	```
	
	Show route removed:

	``` console
	$ ip r | grep "dev zt.* scope link"
	$
	```

	Show container goes unhealthy:

	```
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED          STATUS                     PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   12 minutes ago   Up 12 minutes (unhealthy)             zerotier
	```
	
	Show agent unaware of problem:
	
	```
	$ docker exec zerotier zerotier-cli get 9999888877776666 status
	OK
	```
	
3. Restart container:

	``` console
	$ docker compose restart zerotier
	[+] Restarting 1/1
	 ✔ Container zerotier  Started                                                                                                                             1.5s 
	```
	
	Show route restored:
	
	``` console
	$ ip r | grep "dev zt.* scope link"
	10.244.0.0/16 dev ztr2qsmswx proto kernel scope link src 10.244.235.233 
	```
	
	Show container healthy:
	
	``` console
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED          STATUS                   PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   13 minutes ago   Up 9 seconds (healthy)             zerotier
	```

## Test 4 - join second network

1. Join network:

	``` console
	$ docker exec zerotier zerotier-cli join 9999888877775555
	200 join OK
	```

	Show container goes unhealthy:
	
	```
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED          STATUS                     PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   17 minutes ago   Up 4 minutes (unhealthy)             zerotier
	```
	
	Explore reason:
	
	```
	$ docker exec zerotier zerotier-cli get 9999888877775555 status
	ACCESS_DENIED
	```

	Authorise client in ZeroTier Central. Then show container goes healthy.
	
	``` console
	CONTAINER ID   IMAGE                      COMMAND              CREATED          STATUS                   PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   18 minutes ago   Up 5 minutes (healthy)             zerotier
	```

## Test 5 - interrupt one network

1. List networks:

	```
	$ ip r | grep "dev zt.* scope link"
	10.242.0.0/16 dev ztc3qzoglu proto kernel scope link src 10.242.235.233 
	10.244.0.0/16 dev ztr2qsmswx proto kernel scope link src 10.244.235.233 
	```

2. Destroy one of the networks (would not matter which one):

	``` console
	$ sudo nmcli conn down ztr2qsmswx
	Connection 'ztr2qsmswx' successfully deactivated (D-Bus active path: /org/freedesktop/NetworkManager/ActiveConnection/145)
	```
	
	Show route removed:

	``` console
	$ ip r | grep "dev zt.* scope link"
	10.242.0.0/16 dev ztc3qzoglu proto kernel scope link src 10.242.235.233 
	$
	```

	Show container goes unhealthy:

	```
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED          STATUS                      PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   25 minutes ago   Up 12 minutes (unhealthy)             zerotier
	```
	
	Show agent unaware of problem:
	
	```
	$ docker exec zerotier zerotier-cli listnetworks
	200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
	200 listnetworks 9999888877776666 My_ZeroTier 22:ef:5f:10:91:a9 OK PRIVATE ztr2qsmswx 10.244.235.233/16
	200 listnetworks 9999888877775555 Test 5e:d2:83:cc:ff:c4 OK PRIVATE ztc3qzoglu 10.242.235.233/16
	```

3. Restart container:

	``` console
	$ docker compose restart zerotier
	[+] Restarting 1/1
	 ✔ Container zerotier  Started                                                                                                                             1.5s 
	```
	
	Show route restored:
	
	``` console
	$ ip r | grep "dev zt.* scope link"
	10.242.0.0/16 dev ztc3qzoglu proto kernel scope link src 10.242.235.233 
	10.244.0.0/16 dev ztr2qsmswx proto kernel scope link src 10.244.235.233 
	```
	
	Show container healthy:
	
	``` console
	$ docker ps
	CONTAINER ID   IMAGE                      COMMAND              CREATED          STATUS                    PORTS     NAMES
	0e5b59715ff3   zyclonite/zerotier:local   "entrypoint.sh -U"   27 minutes ago   Up 23 seconds (healthy)             zerotier
	```
	
## Test 5 - down, up the container

```
$ docker compose down zerotier
[+] Running 1/1
 ✔ Container zerotier  Removed                                                                                                                             2.4s 

$ ip r | grep "dev zt.* scope link"

$ docker compose up -d zerotier
[+] Running 1/1
 ✔ Container zerotier  Started                                                                                                                             0.2s 

$ ip r | grep "dev zt.* scope link"
10.242.0.0/16 dev ztc3qzoglu proto kernel scope link src 10.242.235.233 
10.244.0.0/16 dev ztr2qsmswx proto kernel scope link src 10.244.235.233 

$ docker exec zerotier zerotier-cli listnetworks
200 listnetworks <nwid> <name> <mac> <status> <type> <dev> <ZT assigned ips>
200 listnetworks 9999888877776666 My_ZeroTier 22:ef:5f:10:91:a9 OK PRIVATE ztr2qsmswx 10.244.235.233/16
200 listnetworks 9999888877775555 Test 5e:d2:83:cc:ff:c4 OK PRIVATE ztc3qzoglu 10.242.235.233/16

$ docker ps
CONTAINER ID   IMAGE                      COMMAND              CREATED         STATUS                   PORTS     NAMES
c2111263a821   zyclonite/zerotier:local   "entrypoint.sh -U"   8 seconds ago   Up 8 seconds (healthy)             zerotier
```
